File: sample/stream_reader.py
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Callable, Iterator, Optional

from ledger.event_types import BaseEvent, EventType


class StreamReader:
    """Real-time event stream reader for the ledger system."""
    
    def __init__(self, ledger_manager):
        self.ledger_manager = ledger_manager
        self._subscribers: dict = {}
    
    def subscribe(self, event_type: EventType, callback: Callable[[dict], None]) -> None:
        """Subscribe to events of a specific type."""
        if event_type not in self._subscribers:
            self._subscribers[event_type] = []
        self._subscribers[event_type].append(callback)
    
    def unsubscribe(self, event_type: EventType, callback: Callable[[dict], None]) -> None:
        """Unsubscribe from events."""
        if event_type in self._subscribers:
            try:
                self._subscribers[event_type].remove(callback)
            except ValueError:
                pass
    
    def notify_subscribers(self, event: dict) -> None:
        """Notify all subscribers of a new event."""
        # handle dict or object with event_type attribute
        event_type = event.get("event_type") if isinstance(event, dict) else getattr(event, "event_type", None)
        if event_type and event_type in self._subscribers:
            for callback in self._subscribers[event_type]:
                try:
                    callback(event)
                except Exception:
                    pass  # Don't let subscriber errors break the system
    
    def tail_events(self, event_type: EventType, follow: bool = True) -> Iterator[dict]:
        """Tail events from a ledger file, optionally following new events."""
        ledger_file = self.ledger_manager._get_ledger_file(event_type)
        
        if not ledger_file.exists():
            return
            
        with open(ledger_file, 'r') as f:
            # Read existing events
            for line in f:
                try:
                    event_data = json.loads(line.strip())
                    yield event_data
                except json.JSONDecodeError:
                    continue
            
            # Follow new events if requested
            if follow:
                while True:
                    line = f.readline()
                    if line:
                        try:
                            event_data = json.loads(line.strip())
                            yield event_data
                        except json.JSONDecodeError:
                            continue

File: simple-utilities/src/main/java/tech/rsqn/useful/things/ledger/StreamReader.java
package tech.rsqn.useful.things.ledger;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

/**
 * Real-time record stream reader for the ledger system.
 */
public class StreamReader {
    private final LedgerRegistry registry;
    private final Map<RecordType, List<Consumer<Record>>> subscribers = new ConcurrentHashMap<>();

    public StreamReader(LedgerRegistry registry) {
        this.registry = registry;
    }

    public void subscribe(RecordType recordType, Consumer<Record> callback) {
        subscribers.computeIfAbsent(recordType, k -> {
            List<Consumer<Record>> list = Collections.synchronizedList(new ArrayList<>());
            // Subscribe to the underlying ledger once per type, and dispatch to all local subscribers
            Ledger<Record> ledger = (Ledger<Record>) registry.getLedger(recordType);
            if (ledger != null) {
                // We cast to Ledger<Record> to subscribe with a generic Consumer<Record>
                // This works because any T extends Record, so Consumer<Record> can consume it.
                ledger.subscribe(this::notifySubscribers, null);
            }
            return list;
        }).add(callback);
    }

    public void unsubscribe(RecordType recordType, Consumer<Record> callback) {
        List<Consumer<Record>> list = subscribers.get(recordType);
        if (list != null) {
            list.remove(callback);
            if (list.isEmpty()) {
                subscribers.remove(recordType);
                // Note: We don't unsubscribe from the underlying ledger because Ledger interface doesn't support unsubscribe yet
            }
        }
    }

    private void notifySubscribers(Record record) {
        List<Consumer<Record>> list = subscribers.get(record.getType());
        if (list == null || list.isEmpty()) return;
        
        synchronized (list) {
            for (Consumer<Record> callback : list) {
                try {
                    callback.accept(record);
                } catch (Exception e) {
                    // Ignore subscriber errors
                }
            }
        }
    }

    /**
     * Tails records from a ledger, optionally following new records.
     * Uses a callback to process records.
     * 
     * @param recordType The record type to tail.
     * @param follow    Whether to follow new records (tail -f).
     * @param callback  Callback for each record. Returns true to continue, false to stop.
     */
    public void tailEvents(RecordType recordType, boolean follow, ReadCallback<Record> callback) {
        Ledger<Record> ledger = (Ledger<Record>) registry.getLedger(recordType);
        if (ledger == null) {
            return;
        }

        if (!follow) {
            ledger.read(-1, null, callback);
            return;
        }

        // For following, we need to subscribe first to capture records while reading history
        
        // Queue for live records
        java.util.concurrent.BlockingQueue<Record> liveQueue = new java.util.concurrent.LinkedBlockingQueue<>();
        Consumer<Record> listener = liveQueue::offer;
        
        // Subscribe first
        ledger.subscribe(listener, null);
        
        // Track max ID
        java.util.concurrent.atomic.AtomicLong maxId = new java.util.concurrent.atomic.AtomicLong(-1);
        
        // Read history
        // We wrap the callback to update maxId
        boolean[] keepGoing = {true};
        ledger.read(-1, null, (ReadCallback<Record>) record -> {
            if (record.getSequenceId() != null) {
                maxId.accumulateAndGet(record.getSequenceId(), Math::max);
            }
            boolean result = callback.onRecord(record);
            keepGoing[0] = result;
            return result;
        });
        
        if (!keepGoing[0]) {
            // User stopped during history
            return;
        }
        
        // Process live records
        try {
            while (true) {
                Record record = liveQueue.take();
                if (record.getSequenceId() != null && record.getSequenceId() <= maxId.get()) {
                    continue; // Skip duplicate
                }
                
                if (!callback.onRecord(record)) {
                    break;
                }
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
